### PR TITLE
Add ability to define custom cluster

### PR DESCRIPTION
**Pull request by SerhiiTsybulskyi**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

Screenshot 2024-11-21 at 13 50 18

### CODE REVIEW

File: src/GraphScene.tsx [hunk 1]
```diff
@@ -159,6 +160,11 @@ export interface GraphSceneProps {
    */
   renderNode?: NodeRenderer;
 
+  /**
+   * Render a custom cluster label
+   */
+  renderClusterLabel?: ClusterLabelRenderer;
```
Comment (amcdnl) [inline src/GraphScene.tsx:166]: We should just make this render the whole cluster and let the user override whatever they want imo

**Review by amcdnl — CHANGES_REQUESTED** [review]

File: src/GraphScene.tsx (hunk 1 above)
Comment (SerhiiTsybulskyi) [reply src/GraphScene.tsx]: updated

File: src/types.ts [hunk 2]
```diff
@@ -266,13 +267,18 @@ export interface NodeRendererProps {
...
-  label: string;
-  opacity?: number;
-  fontUrl?: string;
+export interface ClusterRendererProps {
+  offset: number;
+  rad: number;
+  padding: number;
+  opacity: SpringValue<number>;
+  label: {
+    position: SpringValue<[number, number, number]>;
```
Comment (amcdnl) [inline src/types.ts:276]: I don't love how `position` is coupled to `SpringValue` like this. No way to just pass number?

File: src/types.ts [hunk 3]
```diff
@@ -266,13 +267,18 @@ export interface NodeRendererProps {
 
 export type NodeRenderer = (args: NodeRendererProps) => ReactNode;
 
-export interface ClusterLabelRendererProps {
-  label: string;
-  opacity?: number;
-  fontUrl?: string;
+export interface ClusterRendererProps {
+  offset: number;
```
Comment (amcdnl) [inline src/types.ts:271]: Needs comments

File: src/symbols/Cluster.tsx [hunk 4]
```diff
@@ -238,54 +240,63 @@ export const Cluster: FC<ClusterProps> = ({
...
+                  attach="material"
+                  color={normalizedFill}
+                  transparent={true}
+                  depthTest={false}
+                  opacity={theme.cluster?.fill ? circleOpacity : 0}
+                  side={DoubleSide}
+                  fog={true}
                 />
+              </mesh>
+              <mesh>
```
Comment (amcdnl) [inline src/symbols/Cluster.tsx:271]: Similar to what we did with `Node` and its pre-defined shapes like `Sphere`, if I were you I would look and see if we could abstract these for a user making it easier to implement this w/o knowing Three.

File: docs/demos/Cluster.story.tsx [hunk 5]
```diff
@@ -467,18 +467,40 @@ export const CustomLabel = () => (
...
+            attach="material"
+            color="gray"
+            transparent={true}
+            depthTest={false}
+            opacity={opacity}
+            side={DoubleSide}
+            fog={true}
+          />
+        </mesh>
+        <a.group position={label.position}>
```
Comment (amcdnl) [inline docs/demos/Cluster.story.tsx:490]: Couldn't this use `Label` directly?

File: src/types.ts (hunk 2 above)
Comment (SerhiiTsybulskyi) [reply src/types.ts:276]: I do it to get rid casting to any as it was previously

Screenshot 2024-11-21 at 17 05 05

Comment (SerhiiTsybulskyi) [reply src/types.ts:276]: I can do it
```
position: SpringValue<[number, number, number]> | [number, number, number] ;
```

File: docs/demos/Cluster.story.tsx (hunk 5 above)
Comment (SerhiiTsybulskyi) [reply docs/demos/Cluster.story.tsx:490]: I've added SVG icon for label, that why it not just Label component. Maybe need to extend label component to define start/end icon (optionally)

File: src/types.ts (hunk 2 above)
Comment (amcdnl) [reply src/types.ts]: I guess I'm trying to say we shouldn't pass a SpringValue, it will cause a ton of re-renders.

Comment (SerhiiTsybulskyi) [reply src/types.ts]: But how it will work if animation is enabled?

File: src/symbols/clusters/Ring.tsx [hunk 6]
```diff
@@ -20,36 +22,48 @@ export const Ring = ({
...
-        transparent={true}
-        depthTest={false}
-        opacity={circleOpacity}
-        side={DoubleSide}
-        fog={true}
-      />
-    </mesh>
-  </>
-);
+}: RingProps) => {
```
Comment (amcdnl) [inline src/symbols/clusters/Ring.tsx:28]: Let's use `FC` instead of this.

**Review by amcdnl — APPROVED** [review]
LGTM other than the fc

**Pull request merged**